### ISSUE TITLE
how to get the final test accuracy

### ISSUE THREAD

**Issue by louislee711** [RenqiChen/Genomic-Selection#8]
I've been reviewing the code, and I noticed that the data is split into train_d and test_d
`train_d, test_d = split_dataset(data_root, fam_path, which_k)`

 I'm curious about which specific dataset was used to calculate the accuracy reported in the paper.
Could you clarify whether the accuracy mentioned in the paper is based on train_d, test_d, or some other evaluation set?

Thanks.

**Comment by Asadiry** [comment]
Thank you for your attention to our work. In calculating the evaluation metrics, we employ the standard k-fold cross-validation procedure: the original dataset is divided into k subsets, with each subset serving as the test set for the calculation of relevant metrics, while the remaining k-1 subsets are used as the training set for model training. The final evaluation metrics are obtained by averaging the results across all folds.

**Comment by louislee711** [comment]
HI
May I ask which part of code are calculating the final evaluation metrics?
Thanks.

**Comment by RenqiChen** [comment]
In `main_rice3k.py`:
```
def test_one_epoch(i_epoch, loader, model, criterion, args):
    print(f"test epoch [{i_epoch}/{args.epoch}]")

    correct_results = {}
    total_results = {}
    total_loss = 0.0

    model.eval()
    for step, data in enumerate(loader):
        x = data["input"] 
        label = data["label"]

        if args.cuda:
            x = x.cuda()
            label = label.cuda()
        
        with torch.no_grad():
            output = model(x)
            loss = criterion(output.squeeze(-1), label)
            total_loss += loss.detach().cpu().item()

            predicted_labels = torch.argmax(output, dim=1)
            correct_predictions = (predicted_labels == label).float() 
            for class_label in torch.unique(label):
                class_mask = (label == class_label).float()
                class_correct = (correct_predictions * class_mask).sum().item()
                class_total = class_mask.sum().item()
                
                if class_label.item() not in correct_results:
                    correct_results[class_label.item()] = 0 
                if class_label.item() not in total_results:
                    total_results[class_label.item()] = 0 

                correct_results[class_label.item()] += class_correct
                total_results[class_label.item()] += class_total

    total_loss /= len(loader)

    total_acc = sum(correct_results.values()) / sum(total_results.values())
    # for key in total_results.keys():
    #     print("class {} accuracy: {:.2f}".format(key, correct_results[key] / total_results[key]))
    #     logger.info("class {} accuracy: {:.2f}".format(key, correct_results[key] / total_results[key]))        

    print("total accuracy: {:.2f}".format(total_acc))
    metric_value = total_acc

    print(f"test epoch[{i_epoch}/{args.epoch}] test loss:{total_loss}")

    return {"loss": total_loss, "accuracy": metric_value}
```
You can see the final evaluation metrics.

**Comment by louislee711** [comment]
Thank you for your reply, I got the information I needed

**Comment by louislee711** [comment]
Thank you for your response, it provided the information I needed.
However, I have a few more questions. Since we are using the test dataset as the final evaluation metric, how can we best avoid overfitting? 
During training with the Rice3K model, I achieved an accuracy of 0.6 at epoch 39, but the final accuracy dropped to 0.48. Any advice on this discrepancy would be greatly appreciated.
thanks

**Comment by RenqiChen** [comment]
Well, overfitting is the most critical issue in these two datasets. Both our approaches are aimed at addressing this issue. This is also an open question for you to explore. A kind piece of advice is to reduce the feature dimensionality.

**Comment by louislee711** [comment]
Thank you for your reply and advice. Have a nice day.
